token-swap/program/src/curve/calculator.rs

```diff
@@ -399,4 +375,57 @@ pub mod test {
                 >= swap_token_b_amount * new_pool_token_supply
         );
     }
+
+    /// Test function checking that a withdraw never reduces the value of pool
+    /// tokens.
+    ///
+    /// Since curve calculations use unsigned integers, there is potential for
+    /// truncation at some point, meaning a potential for value to be lost if
+    /// too much is given to the depositor.
+    pub fn check_pool_value_from_withdraw(
+        curve: &dyn CurveCalculator,
+        pool_token_amount: u128,
+        pool_token_supply: u128,
+        swap_token_a_amount: u128,
+        swap_token_b_amount: u128,
+    ) {
+        let withdraw_result = curve
+            .pool_tokens_to_trading_tokens(
+                pool_token_amount,
+                pool_token_supply,
+                swap_token_a_amount,
+                swap_token_b_amount,
+                RoundDirection::Floor,
+            )
+            .unwrap();
+        let new_swap_token_a_amount = swap_token_a_amount - withdraw_result.token_a_amount;
+        let new_swap_token_b_amount = swap_token_b_amount - withdraw_result.token_b_amount;
+        let new_pool_token_supply = pool_token_supply - pool_token_amount;
+
+        let pool_value = curve
+            .normalized_value(swap_token_a_amount, swap_token_b_amount)
+            .unwrap();
+        // since we can get rounding issues on the pool value which make it seem that the
```

Comment:
Is this because of the divide by 2? I was wondering if you could remove that and have the normalized value be the full total and then do the divide by 2 inside `pool_tokens_to_trading_tokens`. (I'm looking specifically at the constant price curve)

Reply:
It's simply that we can get a situation where our calculation cuts off value from rounding, making the value per token look smaller.  For example, after a withdraw, if the normalized value goes from 100.6 to 100.4 in absolute terms, then our calculation will say it went from 101 to 100, showing that too much value per token was lost.

Reply:
Do you think it's worth changing how the normalized value is created to avoid this? It looks doable for constant price at least

Reply:
We could have the normalized value returned as a `PreciseNumber` instead of an int, otherwise you have to round at some point.  I didn't want to expose that type in the Curve interface, but there shouldn't be a problem with exposing it.

Constant price is actually the only curve that doesn't need this epsilon, which is why the test is implemented differently for it.

Reply:
> We could have the normalized value returned as a PreciseNumber instead of an int

Ah yeah that works too! 

> Constant price is actually the only curve that doesn't need this epsilon, which is why the test is implemented differently for it.

Oh, didn't realize the difference, thanks

Reply:
Done with `fe05a51`